### DISCUSSION TITLE
SerialUPDI guide to follow?

### DISCUSSION THREAD

**Discussion by nabelekt** [SpenceKonde/DxCore#198] [Q&A]
@SpenceKonde, I thought I saw that you moved from trying to maintain multiple guides for creating a SerialUPDI device to having just one, but I still see multiple:

- https://github.com/SpenceKonde/DxCore/blob/master/megaavr/extras/Ref_SerialUPDI.md
- https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md
- https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/tools/README.md
- list may not be exhaustive

**Which is the best to follow?**

From the megaTinyCore version on the CH340:

> Build quality of most boards available online ranges from mediocre to poor. Gross design flaws are often present. I am planning to bring a superior option to marke - it'll have voltage switching done right and a micro USB connector instead of a stuipid full sized USB-A, full-sized 1117 3.3v regulator, and a color-coded power light to indicate the current voltage setting, and a general absence of flagrant design flaws.

**Is that still your plan? Does it exist already?**

Thanks!

**Comment by nabelekt** [comment]
Diffing them, it seems that the DxCore version is the one to use, though none of them have all of the content of them all taken together.

**Comment by SpenceKonde** [comment]
Yes. 
I have had a stack of 144 copies of the bare PCB (panels of 12) that I haven't had time to assemble, it's in the pile with the dozen+ new products that I can't sell yet because I need to take pictures and write up descriptions. 

Assuming that version works well, while some of those will be assembled and sold, hopefully, I'm planning to get them manufactured in quantity,

**Reply by nabelekt** [reply]
If you remember to post here when you have them available, I'd likely get one or two. In the meantime, I'll put my own together.

**Comment by SpenceKonde** [comment]
I really should move everything into the AVR-guidance one.

**Reply by ArduinoShop** [reply]
I recently received my AVR128DB28 chips and went to program them.. but I only had some cheap CP2102s and a decent FTDI with breakout pins for ISP - dunno where they would be used.

The CP2102s had LEDs hanging directly off of the TX/Rx lines - so I put them aside.  I tried playing with diode resistor combos on the FTDI which had no resistors.  Had no joy although I need to RTFM some more in the Guidance README.  Have to admit is tends to ramble a bit and is not easy to follow...

Long story short I put jtag2updi on a proMINI and managed to program the DXcore bootloader first try...  very fast and no pain..  BUT I was unable to upload a program with jtag2updi..  I hooked up my CP2102 to PA5/4 and had a blinking LED in short order but no serial output. Debugging this issue pointed to some issues with DXcore documentation...

1) Spence's famous coloured Pinout diagram seems to have 2 UART0s and 1 UART2..  no UART1 (Tx1/Rx1).
2) this confusion is exacerbated by the option in the IDE tools menu to specify the bootloader upload port which gives the impression that USART0 is on PA4/PA5 which is actually USART1 as far as I can tell.  What I found is that using Serial from the IDE outputs on PA0/PA1.
Serial1 does not seem to output anywhere - which begs the question - Does the bootloader prevent the IDE from using Serial1??  I had no problem with Serial2 on PF0/PF1.

Anyway - on to taking a stab at the RTC feature..

Let's all thank Spence for his hard work...

**Comment by SpenceKonde** [comment]
https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md is the correct reference to use. I'll try correct all theother documents to point there, as it has proven hopeless to keep them all in sync.

**Comment by SpenceKonde** [comment]
Okay, at some point previously I'd made both the other pages link to the unified one.
